src/common/lib/transport/connectionmanager.ts

```diff
@@ -1508,129 +1261,160 @@ class ConnectionManager extends EventEmitter {
+              'http connectivity check succeeded'
+            );
+          }
+        });
+      }
+    }, 4_000);
+  }
 
-    if (!this.transports.includes(preference)) {
-      this.unpersistTransportPreference();
-      this.connectImpl(transportParams, connectCount);
+  cancelWebSocketSlowTimer() {
+    if (this.webSocketSlowTimer) {
+      clearTimeout(this.webSocketSlowTimer);
+      this.webSocketSlowTimer = null;
     }
+  }
 
-    Logger.logAction(
-      Logger.LOG_MINOR,
-      'ConnectionManager.connectPreference()',
-      'Trying to connect with stored transport preference ' + preference
-    );
-
-    const preferenceTimeout = setTimeout(() => {
-      preferenceTimeoutExpired = true;
-      if (!(this.state.state === this.states.connected.state)) {
+  startWebSocketGiveUpTimer(transportParams: TransportParams) {
+    this.webSocketGiveUpTimer = setTimeout(() => {
+      if (!this.wsCheckResult) {
+        this.abandonedWebSocket = true;
+        // if we don't have a base transport to fallback to, just let the websocket connection attempt time out
         Logger.logAction(
           Logger.LOG_MINOR,
-          'ConnectionManager.connectPreference()',
-          'Shortcircuit connection attempt with ' + preference + ' failed; clearing preference and trying from scratch'
+          'ConnectionManager WebSocket give up timer',
+          'websocket connection took more than 10s; ' + (this.baseTransport ? 'trying base transport' : '')
         );
-        /* Abort all connection attempts. (This also disconnects the active
-         * protocol, but none exists if we're not in the connected state) */
-        this.disconnectAllTransports();
-        /* Be quite agressive about clearing the stored preference if ever it doesn't work */
-        this.unpersistTransportPreference();
-      }
-      this.connectImpl(transportParams, connectCount);
-    }, this.options.timeouts.preferenceConnectTimeout);
-
-    /* For connectPreference, just use the main host. If host fallback is needed, do it in connectBase.
-     * The wstransport it will substitute the httphost for an appropriate wshost */
-    transportParams.host = this.httpHosts[0];
-    this.tryATransport(transportParams, preference, (fatal: boolean, transport: Transport) => {
-      clearTimeout(preferenceTimeout);
-      if (preferenceTimeoutExpired && transport) {
-        /* Viable, but too late - connectImpl() will already be trying
-         * connectBase, and we weren't in upgrade mode. Just remove the
-         * onconnected listener and get rid of it */
-        transport.off();
-        transport.disconnect();
-        Utils.arrDeleteValue(this.pendingTransports, transport);
-      } else if (!transport && !fatal) {
-        /* Preference failed in a transport-specific way. Try more */
-        this.unpersistTransportPreference();
-        this.connectImpl(transportParams, connectCount);
+        if (this.baseTransport) {
+          this.connectBase(transportParams);
```

Comment:
What happens if the websocket succeeds after we kick off connectBase? I guess whatever one succeeds second will just activate over the top of the first one? that doesn't seem ideal.

Why not cancel it here so you avoid the chance of both transports succeeding? I guess the answer is that you've removed the proposedTrasnports mechanism so you don't actually have a way to cancel it, what's the reasoning behind doing that?

Reply:
I've re-implemented `connectCounter` and added a `proposedTransport` (which, if set, is disposed when we fallback from ws to base) so this should be solid now